File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/library/items/playlists/playback/PlaylistPlayer.java
package com.lasthopesoftware.bluewater.client.library.items.playlists.playback;

import com.lasthopesoftware.bluewater.client.library.items.media.files.playback.file.IPlaybackHandler;
import com.lasthopesoftware.bluewater.client.library.items.media.files.playback.file.PositionedPlaybackFile;
import com.lasthopesoftware.bluewater.client.library.items.media.files.playback.file.preparation.queues.IPreparedPlaybackFileQueue;
import com.lasthopesoftware.promises.IPromise;
import com.vedsoft.futures.callables.VoidFunc;

import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Closeable;
import java.io.IOException;

import io.reactivex.ObservableEmitter;

/**
 * Created by david on 11/8/16.
 */
public final class PlaylistPlayer implements IPlaylistPlayer, Closeable {

	private static final Logger logger = LoggerFactory.getLogger(PlaylistPlayer.class);
	private final IPreparedPlaybackFileQueue preparedPlaybackFileProvider;
	private final int preparedPosition;
	private PositionedPlaybackFile positionedPlaybackFile;
	private float volume;

	private volatile boolean isStarted;
	private ObservableEmitter<PositionedPlaybackFile> emitter;

	public PlaylistPlayer(@NotNull IPreparedPlaybackFileQueue preparedPlaybackFileProvider, int preparedPosition) {
		this.preparedPlaybackFileProvider = preparedPlaybackFileProvider;
		this.preparedPosition = preparedPosition;
	}

	@Override
	public void subscribe(ObservableEmitter<PositionedPlaybackFile> e) throws Exception {
		emitter = e;

		if (!isStarted) {
			isStarted = true;
			setupNextPreparedFile(preparedPosition);
		}
	}

	public void pause() {
		if (positionedPlaybackFile == null) return;

		final IPlaybackHandler playbackHandler = positionedPlaybackFile.getPlaybackHandler();

		if (playbackHandler.isPlaying()) playbackHandler.pause();
	}

	@Override
	public void resume() {
		if (positionedPlaybackFile != null && !positionedPlaybackFile.getPlaybackHandler().isPlaying())
			positionedPlaybackFile.getPlaybackHandler().promisePlayback();
	}

	@Override
	public void setVolume(float volume) {
		this.volume = volume;

		final PositionedPlaybackFile positionedPlaybackFile = this.positionedPlaybackFile;
		if (positionedPlaybackFile != null)
			positionedPlaybackFile.getPlaybackHandler().setVolume(volume);
	}

	private void setupNextPreparedFile() {
		setupNextPreparedFile(0);
	}

	private void setupNextPreparedFile(int preparedPosition) {
		final IPromise<PositionedPlaybackFile> preparingPlaybackFile =
			preparedPlaybackFileProvider
				.promiseNextPreparedPlaybackFile(preparedPosition);

		if (preparingPlaybackFile == null) {
			doCompletion();
			return;
		}

		preparingPlaybackFile
			.then(this::changePlaybackFile)
			.thenPromise(this::startFilePlayback)
			.error(VoidFunc.running(this::handlePlaybackException));
	}

	private PositionedPlaybackFile changePlaybackFile(@NotNull PositionedPlaybackFile positionedPlaybackFile) {
		this.positionedPlaybackFile = positionedPlaybackFile;

		emitter.onNext(this.positionedPlaybackFile);

		return positionedPlaybackFile;
	}

	private IPromise<IPlaybackHandler> startFilePlayback(@NotNull PositionedPlaybackFile positionedPlaybackFile) {
		final IPlaybackHandler playbackHandler = positionedPlaybackFile.getPlaybackHandler();

		playbackHandler.setVolume(volume);

		final IPromise<IPlaybackHandler> promisedPlayback = playbackHandler.promisePlayback();

		promisedPlayback
			.then(VoidFunc.running(this::closeAndStartNextFile))
			.error(VoidFunc.running(this::handlePlaybackException));

		return promisedPlayback;
	}

	private void closeAndStartNextFile(IPlaybackHandler playbackHandler) {
		try {
			playbackHandler.close();
		} catch (IOException e) {
			logger.error("There was an error releasing the media player", e);
		}

		setupNextPreparedFile();
	}

	private void haltPlayback() {
		try {
			if (positionedPlaybackFile != null)
				positionedPlaybackFile.getPlaybackHandler().close();

			doCompletion();
		} catch (IOException e) {
			logger.error("There was an error releasing the media player", e);
			emitter.onError(e);
		}
	}

	private void handlePlaybackException(Exception exception) {
		emitter.onError(exception);

		haltPlayback();
	}

	@Override
	public void close() throws IOException {
		haltPlayback();
	}

	private void doCompletion() {
		this.positionedPlaybackFile = null;

		emitter.onComplete();
	}
}

File: playback/src/main/java/com/namehillsoftware/projectblue/playback/playlist/PlaylistPlayer.kt
package com.namehillsoftware.projectblue.playback.playlist

import com.lasthopesoftware.bluewater.client.playback.playlist.IPlaylistPlayer
import com.namehillsoftware.handoff.promises.Promise
import com.namehillsoftware.projectblue.playback.engine.preparation.PreparedPlayableFileQueue
import com.namehillsoftware.projectblue.playback.file.PlayableFile
import com.namehillsoftware.projectblue.playback.file.PlayingFile
import com.namehillsoftware.projectblue.playback.file.PositionedPlayableFile
import com.namehillsoftware.projectblue.playback.file.PositionedPlayingFile
import com.namehillsoftware.projectblue.shared.promises.extensions.toPromise
import io.reactivex.ObservableEmitter
import org.slf4j.LoggerFactory
import java.io.Closeable
import java.io.IOException

class PlaylistPlayer(private val preparedPlaybackFileProvider: PreparedPlayableFileQueue, private val preparedPosition: Long) : IPlaylistPlayer, Closeable {

	companion object {
		private val logger = LoggerFactory.getLogger(PlaylistPlayer::class.java)
	}

	private val stateChangeSync = Any()
	private var positionedPlayingFile: PositionedPlayingFile? = null
	private var positionedPlayableFile: PositionedPlayableFile? = null
	private var volume = 0f

	private var lastStateChangePromise: Promise<*> = Promise.empty<Any>()

	@Volatile
	private var isStarted = false
	private var emitter: ObservableEmitter<PositionedPlayingFile>? = null

	override fun subscribe(e: ObservableEmitter<PositionedPlayingFile>) {
		emitter = e
		if (!isStarted) {
			isStarted = true
			setupNextPreparedFile(preparedPosition)
		}
	}

	override fun pause(): Promise<PositionedPlayableFile?> {
		synchronized(stateChangeSync) {
			return lastStateChangePromise
				.eventually({ promisePause() }, { promisePause() })
				.also { lastStateChangePromise = it }
		}
	}

	override fun resume(): Promise<PositionedPlayingFile?> {
		synchronized(stateChangeSync) {
			return lastStateChangePromise
				.eventually({ promiseResumption() }, { promiseResumption() })
				.also { lastStateChangePromise = it }
		}
	}

	override val isPlaying: Boolean
		get() {
			return positionedPlayingFile != null
		}

	override fun setVolume(volume: Float) {
		this.volume = volume
		positionedPlayableFile?.playableFileVolumeManager?.volume = volume
	}

	private fun promisePause(): Promise<PositionedPlayableFile?> {
		return positionedPlayingFile
			?.let {
				it.playingFile
                        .promisePause()
					?.then { p ->
						PositionedPlayableFile(p, it.playableFileVolumeManager,	it.asPositionedFile()).apply {
							positionedPlayableFile = this
							positionedPlayingFile = null
						}
					}
			}
			?: Promise(IllegalStateException("There must be a playing file in order to pause"))
	}

	private fun promiseResumption(): Promise<PositionedPlayingFile?> {
		return positionedPlayableFile
			?.let {
				it.playableFile
					?.promisePlayback()
					?.then { p ->
						PositionedPlayingFile(p, it.playableFileVolumeManager, it.asPositionedFile()).apply {
							positionedPlayingFile = this
							positionedPlayableFile = null
						}
						positionedPlayingFile
					}
			}
			?: Promise(IllegalStateException("There must be a playable file in order to resume"))
	}

	private fun setupNextPreparedFile(preparedPosition: Long = 0) {
		val preparingPlaybackFile = preparedPlaybackFileProvider
			.promiseNextPreparedPlaybackFile(preparedPosition)

		if (preparingPlaybackFile == null) {
			doCompletion()
			return
		}

		preparingPlaybackFile
			.eventually { positionedPlayableFile -> startFilePlayback(positionedPlayableFile) }
			.excuse { exception -> handlePlaybackException(exception) }
	}

	private fun startFilePlayback(positionedPlayableFile: PositionedPlayableFile): Promise<PlayingFile?> {
		positionedPlayableFile.playableFileVolumeManager.volume = volume
		val playbackHandler = positionedPlayableFile.playableFile

		synchronized(stateChangeSync) {
			this.positionedPlayableFile = positionedPlayableFile
			val promisedPlayback = lastStateChangePromise.eventually(
				{ playbackHandler.promisePlayback() },
				{ playbackHandler.promisePlayback() })

			lastStateChangePromise = promisedPlayback
				.then { playingFile ->
					val newPositionedPlayingFile = PositionedPlayingFile(
						playingFile,
						positionedPlayableFile.playableFileVolumeManager,
						positionedPlayableFile.asPositionedFile())

					positionedPlayingFile = newPositionedPlayingFile

					emitter?.onNext(newPositionedPlayingFile)

					newPositionedPlayingFile
						.playingFile
						.promisePlayedFile()
						.then(
							{ closeAndStartNextFile(playbackHandler) },
							{ handlePlaybackException(it) })
				}
			return promisedPlayback
		}
	}

	private fun closeAndStartNextFile(playbackHandler: PlayableFile) {
		try {
			playbackHandler.close()
		} catch (e: IOException) {
			logger.error("There was an error releasing the media player", e)
		}
		setupNextPreparedFile()
	}

	private fun haltPlayback() {
		synchronized(stateChangeSync) {
			lastStateChangePromise = lastStateChangePromise
				.eventually(
					{ generateHaltPromise() },
					{ generateHaltPromise() })
				.then({ p ->
					p?.close()
					doCompletion()
				}, { e ->
					logger.error("There was an error releasing the media player", e)
					emitter?.onError(e)
				})
		}
	}

	private fun generateHaltPromise(): Promise<PlayableFile> {
		return positionedPlayableFile?.playableFile?.toPromise()
			?: positionedPlayingFile?.playingFile?.promisePause()
			?: Promise.empty()
	}

	private fun handlePlaybackException(exception: Throwable) {
		emitter?.onError(exception)
		haltPlayback()
	}

	override fun close() {
		haltPlayback()
	}

	private fun doCompletion() {
		positionedPlayingFile = null
		emitter?.onComplete()
	}
}
